### ISSUE TITLE
Performances

### ISSUE THREAD

**Issue by Clemv95** [nitrog0d/palworld-arm64#4]
Hi

I used your image on a oracle free tier, with 4 cpu and 24gig of ram
The performances are ok, but around 34 fps ( server fps in game) with multithreading enabled
I assume this is because of FEX, kind of emulating stuff
Do you have a better fps ?

**Comment by nitrog0d** [comment]
That's as good as it'll get lol, running a world for 2 days now, it's around 30 fps. However I see no difference from 60 to 30 server-side.

**Comment by BrunoDomenesDutra** [comment]
> Hi
> 
> I used your image on a oracle free tier, with 4 cpu and 24gig of ram The performances are ok, but around 34 fps ( server fps in game) with multithreading enabled I assume this is because of FEX, kind of emulating stuff Do you have a better fps ?

Its 'normal'. Right now it doesn't matter how robust your server is. It's up to PocketPair to fix this issue.

**Comment by RhanCandia** [comment]
Yeah, I think this comes down to the optimization the developers need to make.
I have the exact same configuration on OCI.

There are 3 active players with really busy bases. Server FPS is about 20 fps now.

Sometimes when I go away from my base for too long, when I get back all my Pals are not moving and are all starving even though there are plenty of resources in there. It's like the AI shuts off when I was away for too long.

**Comment by BrunoDomenesDutra** [comment]
> Yeah, I think this comes down to the optimization the developers need to make. I have the exact same configuration on OCI.
> 
> There are 3 active players with really busy bases. Server FPS is about 20 fps now.
> 
> Sometimes when I go away from my base for too long, when I get back all my Pals are not moving and are all starving even though there are plenty of resources in there. It's like the AI shuts off when I was away for too long.

Have the same problem here regarding the Pals.

One thing I did was to lower the number os player slots from 32 to 10. My friends said after that, the server was better.

**Comment by RhanCandia** [comment]
> > Yeah, I think this comes down to the optimization the developers need to make. I have the exact same configuration on OCI.
> > There are 3 active players with really busy bases. Server FPS is about 20 fps now.
> > Sometimes when I go away from my base for too long, when I get back all my Pals are not moving and are all starving even though there are plenty of resources in there. It's like the AI shuts off when I was away for too long.
> 
> Have the same problem here regarding the Pals.
> 
> One thing I did was to lower the number os player slots from 32 to 10. My friends said after that, the server was better.

I will try that. I’ll let you know after. Thanks for sharing your experience.

**Comment by Clemv95** [comment]
> Yeah, I think this comes down to the optimization the developers need to make. I have the exact same configuration on OCI.
> 
> There are 3 active players with really busy bases. Server FPS is about 20 fps now.
> 
> Sometimes when I go away from my base for too long, when I get back all my Pals are not moving and are all starving even though there are plenty of resources in there. It's like the AI shuts off when I was away for too long.

Same problem here with pals getting stucks

**Comment by nitrog0d** [comment]
Sadly there's nothing I can do about it. I'd say this is PocketPair's fault mostly, the server runs like shit on normal amd64 machines too. And still needs all that ram

**Comment by RhanCandia** [comment]
> > Yeah, I think this comes down to the optimization the developers need to make. I have the exact same configuration on OCI.
> > There are 3 active players with really busy bases. Server FPS is about 20 fps now.
> > Sometimes when I go away from my base for too long, when I get back all my Pals are not moving and are all starving even though there are plenty of resources in there. It's like the AI shuts off when I was away for too long.
> 
> Have the same problem here regarding the Pals.
> 
> One thing I did was to lower the number os player slots from 32 to 10. My friends said after that, the server was better.

Seriously man. That worked! From 32, I changed my max server player to 10. From regularly being at ServerFPS 20, now we are back up at above 50! Thanks!

**Comment by nitrog0d** [comment]
I'm impressed that works lol

**Comment by BrunoDomenesDutra** [comment]
Bad netcode. The server probably pre allocate resources for the player slots and keep them. So we have 'ghost' players on the server eating resource.

**Comment by Clemv95** [comment]
> > Yeah, I think this comes down to the optimization the developers need to make. I have the exact same configuration on OCI.
> > There are 3 active players with really busy bases. Server FPS is about 20 fps now.
> > Sometimes when I go away from my base for too long, when I get back all my Pals are not moving and are all starving even though there are plenty of resources in there. It's like the AI shuts off when I was away for too long.
> 
> Have the same problem here regarding the Pals.
> 
> One thing I did was to lower the number os player slots from 32 to 10. My friends said after that, the server was better.

I did the same thing, lower the max players to 6, but i dont see any improvement

**Comment by Clemv95** [comment]
> > > Yeah, I think this comes down to the optimization the developers need to make. I have the exact same configuration on OCI.
> > > There are 3 active players with really busy bases. Server FPS is about 20 fps now.
> > > Sometimes when I go away from my base for too long, when I get back all my Pals are not moving and are all starving even though there are plenty of resources in there. It's like the AI shuts off when I was away for too long.
> > 
> > 
> > Have the same problem here regarding the Pals.
> > One thing I did was to lower the number os player slots from 32 to 10. My friends said after that, the server was better.
> 
> Seriously man. That worked! From 32, I changed my max server player to 10. From regularly being at ServerFPS 20, now we are back up at above 50! Thanks!

Does it fix the pals not moving when going to far from base ?

**Comment by LicensedTerrapin** [comment]
> > > > Yeah, I think this comes down to the optimization the developers need to make. I have the exact same configuration on OCI.
> > > > There are 3 active players with really busy bases. Server FPS is about 20 fps now.
> > > > Sometimes when I go away from my base for too long, when I get back all my Pals are not moving and are all starving even though there are plenty of resources in there. It's like the AI shuts off when I was away for too long.
> > > 
> > > 
> > > Have the same problem here regarding the Pals.
> > > One thing I did was to lower the number os player slots from 32 to 10. My friends said after that, the server was better.
> > 
> > 
> > Seriously man. That worked! From 32, I changed my max server player to 10. From regularly being at ServerFPS 20, now we are back up at above 50! Thanks!
> 
> Does it fix the pals not moving when going to far from base ?

I haven't noticed pals stopping work when I leave but I started my server with max 10 players anyways. However my crashes after a few hours and docker doesn't restart it so that's a problem too...

**Comment by r3ps4J** [comment]
Do you guys use the multithreading parameter on Oracle Cloud free tier? I noticed that my server performance drops significantly as soon as there is more than 1 player online. With just me the server fps is around 50-60, when the second player joins it drops down to around 30 with a lot more lagg during gameplay. I already reduced the max player count to 4.

**Comment by r3ps4J** [comment]
> > > > > Yeah, I think this comes down to the optimization the developers need to make. I have the exact same configuration on OCI.
> > > > > There are 3 active players with really busy bases. Server FPS is about 20 fps now.
> > > > > Sometimes when I go away from my base for too long, when I get back all my Pals are not moving and are all starving even though there are plenty of resources in there. It's like the AI shuts off when I was away for too long.
> > > > 
> > > > 
> > > > Have the same problem here regarding the Pals.
> > > > One thing I did was to lower the number os player slots from 32 to 10. My friends said after that, the server was better.
> > > 
> > > 
> > > Seriously man. That worked! From 32, I changed my max server player to 10. From regularly being at ServerFPS 20, now we are back up at above 50! Thanks!
> > 
> > 
> > Does it fix the pals not moving when going to far from base ?
> 
> I haven't noticed pals stopping work when I leave but I started my server with max 10 players anyways. However my crashes after a few hours and docker doesn't restart it so that's a problem too...

I use docker compose with restart: unless-stopped. That way it restarts it when it crashes, but I've had it happen that it crashes but doesn't actually exit so docker doesn't know it needs to restart it. Anyway, you can read more about a restart policy here: https://docs.docker.com/config/containers/start-containers-automatically/

**Comment by sonroyaalmerol** [comment]
I've been having issues with performance as well (was having around 15-20fps with 3 players on the server). I'm also on oracle's free tier.

I tried making a container with Box64 instead of FEX for emulation and now I'm getting 40-60fps on the same hardware. No idea what the technical differences between the Box64 and FEX are (it's my first time encountering both) so I can't say it'll do the same for others, but you can try it out: https://github.com/sonroyaalmerol/palworld-server-docker-arm64

**Comment by nitrog0d** [comment]
Test more? It depends on how long the server has been online for and what the players are doing 
That being said, I'll test it! Thank you for the contribution to the community :)

**Comment by LicensedTerrapin** [comment]
> Test more? It depends on how long the server has been online for and what the players are doing That being said, I'll test it! Thank you for the contribution to the community :)

I had a look at some benchmarks however they're quite old: 
[https://box86.org/2022/03/box86-box64-vs-qemu-vs-fex-vs-rosetta2/](url)

According to this box should be faster but a lotta years and development have passed since this was done, also by box86.org so who knows... Currently it does seem like 3 players kinda max out an oracle free tier ampere. :|

Thank you for your hard work mate, appreciate it.

**Comment by r3ps4J** [comment]
I've been running the box64 image for some time now. I noticed that server fps is higher in general, but it's less stable than this FEX image. The fps fluctuates more and pals seem buggier overall. I noticed too that I experienced the pals starving and being depressed bug more often on the box64 image. Not sure what would cause that, maybe because pals got stuck more often there.
